1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+```
+
+```powershell
+Write-Error "Bad" || Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+
+Hello again
+```
+
+Using `echo` as a convenient example of a successful command
+and `false` as an example of an unsuccesful command.
+
+```powershell
+echo "Hello" && Write-Output "Hi"
+```
+
+```output
+Hello
+Hi
+```
+
+```powershell
+false && Write-Output "Reached"
+```
+
+```output
+```
+
+```powershell
+false || Write-Output "Reached"
+```
+
+```output
+Reached
+```
+
+```powershell
+false || Write-Output "Command failed" && Write-Output "Backup"
+```
+
+```output
+Command failed
+Backup
+```
+
+Also see: [intended test cases for implementation](https://github.com/PowerShell/PowerShell/blob/f8b899e42c86957a1b58273be0029f40a67bf1b6/test/powershell/Language/Operators/BashControlOperator.Tests.ps1).
+
+## Specification
+
+### Grammar
+
+Pipeline chains will be implemented using the already reserved `&&` and `||` operators.
+These operators will have the following grammar:
+
+```none
+statement:
+    | ... <existing statements except pipeline>   # Other statements
+    | variable_expression "=" statement           # Assignment
+    | pipeline_chain ["&"]                        # Pipeline chains
+
+pipeline_chain:
+    | pipeline
+    | pipeline_chain "&&" [newlines] pipeline
+    | pipeline_chain "||" [newlines] pipeline
+```
+
+#### Optional pipeline chaining
+
+A pipeline chain is a chain of one or more pipelines
+and takes the place of a pipeline in the current PowerShell syntax.
+
+In the degenerate case, a pipeline chain of a single pipeline
+will work exactly as pipelines currently do in PowerShell,
+so there is no breakage to the existing grammar.
+
+#### Line continuation
+
+After a pipeline chain operator,
+any newlines will be skipped in anticipation of the following pipeline.
+
+For example, the following would be a single pipeline chain:
+
+```powershell
+cmd1 &&
+    cmd2 ||
+    cmd3
+```
+
+If the end of file is reached after a pipeline chain operator,
+incomplete input will be reported so that integrating tools
+will know to keep prompting for input.
+
+#### Left-associativity
+
+Like in POSIX shells, pipeline chain operators will be left associative,
+meaning chain operators will group from left to right.
+
+For example, given the following pipeline chain:
+
+```powershell
+cmd1 && cmd2 || cmd3
+```
+
+This will be grouped as:
+
+```none
+[[cmd1 && cmd2] || cmd3]
+```
+
+As a syntax tree, this would look like:
+
+```none
+            "||"
+          /      \
+        "&&"     "cmd3"
+      /      \
+   "cmd1"    "cmd2"
+```
+
+With the syntax tree deepening on the left as more operators are chained.
+
+Semantically, this means that `cmd1 && cmd2` would be evaluated first,
+and its result used to govern the evaluation of `|| cmd3`.
+

```

Comment:
I think an additional text is needed here for clarity, as follows:

If `cmd1` is successful, the results of `cmd2` will be tested for failure, in which case `cmd3` will be invoked.

If `cmd1` is unsuccessful, `cmd3` will be invoked.